### PR TITLE
MAINT: Fixes from static analysis

### LINKED ISSUE #12665: List of possible bugs found by automated code analysis
I ran an automated code analysis by [https://www.deepcode.ai/](https://www.deepcode.ai/) and the AI found the following bugs. I'm not sure if all of them are serious or important but if two or three code improvements can be made, that's something at least.

1. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/sparse/linalg/dsolve/SuperLU/SRC/mmd.c#L400
 "The expression will always evaluate to true because both sides always hold the same value."
2. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/sparse/linalg/dsolve/SuperLU/SRC/util.c#L423 "This division [:423] may result in a division by zero error because 0 [:406] is a possible divisor."
3. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/spatial/qhull_src/src/random_r.c#L53 "Using strcpy [:53] can lead to severe buffer overflow vulnerabilities. Use the safe alternative strncpy instead."
4. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/benchmarks/benchmarks/signal_filtering.py#L65 "Missing close for concurrent.futures.ThreadPoolExecutor [:65] , add close or use a with block"
5. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/optimize/_minimize.py#L832 "There is no need to call list [:832] on an object [:832] that is already constructed as a list [:824] ."
6. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/pavement.py#L210 "Missing close for open [:210] , add close or use a with block."
7. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/special/_generate_pyx.py#L460 "Use comprehensions instead of map [:460] ."
8. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/optimize/nonlin.py#L964 "Accessing attribute svd_reduce [:964] on possibly None value. None flows [:950, :964] from None literal [:950] . Consider adding a check."
9. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/stats/_binned_statistic.py#L659 "Trying to call len [:659] on an object of type 0 (0) will lead to a TypeError."
10. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/tools/openblas_support.py#L154 "Use os.makedirs instead of os.mkdir [:154] because the given path may require creating the parent directories."
11. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/tools/refguide_check.py#L284 "There is no need to call set [:284] on an object [:284] that is already constructed as a set [:249]"
12. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/ndimage/src/_ctest.c#L47 "Use Py_XDECREF instead of Py_DECREF [:47]"
13. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/optimize/tnc/moduleTNC.c#L71 "Use Py_XDECREF instead of Py_DECREF [:71]"
14. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/special/cephes/jv.c#L91 "Equals comparison of floating point numbers. Floating point numbers are only considered equal if their bit representation is equal. Consider using an epsilon comparison instead."
15. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/signal/sigtoolsmodule.c#L1144 "Numerical operation multiply [:1144] used with boolean operand notequals [:1144] . Consider making the conversion from bool to integer explicit."
16. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/sparse/linalg/dsolve/SuperLU/SRC/zgscon.c#L102 "Comparison of integral operand 0 [:102] and boolean operand not [:102] . Consider making both sides the same type, or check the expression by comparing to true/false/TRUE/FALSE or use if (x), if (!x)."
17. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/sparse/sparsetools/sparsetools.cxx#L488 "Caught exception e [:488] seems to be ignored in catch block [:488] . Consider logging or rethrowing e [:488] ."
18. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/benchmarks/benchmarks/linalg.py#L18 "Use predictable random with seed or secure random."
19. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/linalg/tests/test_decomp_ldl.py#L77 "Use predictable random with seed or secure random."
20. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/optimize/_dual_annealing.py#L410 "No need to call keys [:410] , directly check with the in operator."
21. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/special/tests/test_sf_error.py#L47 "No need to call keys [:47] , directly check with the in operator."
22. https://github.com/scipy/scipy/blob/f7f31b4574c9665afd87068fb466f78a7bc37dd6/scipy/special/_testutils.py#L299 "Use comprehensions instead of map [:299] ."

### PR DESCRIPTION
**Pull request by sethtroisi**
Fixes a handful of static issues found in #12665
closes #12665

### CODE REVIEW

File: benchmarks/benchmarks/signal_filtering.py [hunk 1]
```diff
@@ -62,12 +62,12 @@ def setup(self, n_samples, threads):
...
-        futures = []
-        for i in range(threads):
-            futures.append(pool.submit(sosfilt, self.filt, self.chunks[i]))
+        with ThreadPoolExecutor(max_workers=threads) as pool:
+            futures = []
+            for i in range(threads):
+                futures.append(pool.submit(sosfilt, self.filt, self.chunks[i]))
 
-        wait(futures)
+            wait(futures)
```
Comment (sethtroisi) [inline benchmarks/benchmarks/signal_filtering.py:70]: I tested with `python runtests.py --bench signal.filtering`

File: pavement.py [hunk 2]
```diff
@@ -206,8 +206,9 @@ def release(options):
 def compute_md5(idirs):
     released = paver.path.path(idirs).listdir()
     checksums = []
-    for f in sorted(released):
-        m = md5(open(f, 'rb').read())
+    for fn in sorted(released):
+        with open(fn, 'rb') as f:
+            m = md5(f.read())
```
Comment (sethtroisi) [inline pavement.py:211]: I don't know how to test these.
But the changes are small and easy to comprehend.

File: pavement.py [hunk 3]
```diff
@@ -267,9 +269,8 @@ def write_log_task(filename='Changelog'):
             stdout=subprocess.PIPE)
 
     out = st.communicate()[0].decode()
-    a = open(filename, 'w')
-    a.writelines(out)
-    a.close()
+    with open(filename, 'w') as a:
+        a.writelines(out)
```
Comment (sethtroisi) [inline pavement.py:273]: Can this be simplified to
```
    with open(filename, 'w') as log:
        st = subprocess.Popen(
            ['git', 'log',  '%s..%s' % (LOG_START, LOG_END)],	            ['git', 'log',  '%s..%s' % (LOG_START, LOG_END)],
            stdout=log)

File: pavement.py (hunk 2 above)
Comment (rgommers) [reply pavement.py:211]: You'd test this by installing `paver` and running `paver release`. But no need, this is clearly correct.

File: pavement.py (hunk 3 above)
Comment (rgommers) [reply pavement.py:273]: Maybe, but I'd rather not spend time to test that right now. This has worked for years and is only used by the release manager.

**Review by rgommers — APPROVED** [review]
LGTM, thanks @sethtroisi

**Pull request merged**